File: src/main/java/axirassa/webapp/pages/user/ResetPasswordUser.java

package axirassa.webapp.pages.user;

import org.apache.shiro.authz.annotation.RequiresGuest;
import org.apache.tapestry5.annotations.Component;
import org.apache.tapestry5.annotations.Parameter;
import org.apache.tapestry5.annotations.Property;
import org.apache.tapestry5.annotations.Secure;
import org.apache.tapestry5.hibernate.annotations.CommitAfter;
import org.apache.tapestry5.ioc.annotations.Inject;

import axirassa.dao.UserEmailAddressDAO;
import axirassa.model.UserEntity;
import axirassa.model.flows.ResetPasswordFlow;
import axirassa.webapp.components.AxForm;
import axirassa.webapp.components.AxTextField;

@Secure
@RequiresGuest
public class ResetPasswordUser {
	@Inject
	private UserEmailAddressDAO emailDAO;

	@Inject
	private ResetPasswordFlow resetFlow;

	@Property
	private String email;

	@Parameter("message:unknownEmail-message")
	private String unknownEmailMessage;

	@Component
	private AxForm form;

	@Component
	private AxTextField emailField;


	public void onValidateFromForm() {
		if (email == null) {
			showInvalidEmailMessage();
			return;
		}

		UserEntity entity = emailDAO.getUserByEmail(email);
		if (entity == null)
			showInvalidEmailMessage();
	}


	private void showInvalidEmailMessage() {
		form.recordError(emailField, unknownEmailMessage);
	}


	@CommitAfter
	public Object onSuccessFromForm() throws Exception {
		UserEntity user = emailDAO.getUserByEmail(email);
		resetFlow.setUserEntity(user);
		resetFlow.execute();

		return PasswordResetSentUser.class;
	}
}

File: src/main/java/axirassa/webapp/pages/user/ResetPasswordUser.scala
package axirassa.webapp.pages.user

import axirassa.dao.UserDAO
import axirassa.model.PasswordResetTokenEntity
import axirassa.services.email.EmailTemplate
import axirassa.webapp.components.AxForm
import axirassa.webapp.services.EmailNotifyService
import org.apache.shiro.authz.annotation.RequiresGuest
import org.apache.tapestry5.annotations.Component
import org.apache.tapestry5.annotations.Property
import org.apache.tapestry5.annotations.Secure
import org.apache.tapestry5.hibernate.annotations.CommitAfter
import org.apache.tapestry5.ioc.annotations.Inject
import org.apache.tapestry5.services.PageRenderLinkSource
import org.hibernate.Session
@Secure
@RequiresGuest
class ResetPasswordUser {

    @Inject
    var database : Session = _

    @Inject
    var userDAO : UserDAO = _

    @Inject
    var linkSource : PageRenderLinkSource = _

    @Inject
    var emailNotify : EmailNotifyService = _

    @Property
    var email : String = _

    @Component
    var form : AxForm = _

    def onValidateFromForm() {
        if (email == null) {
            showInvalidEmailMessage()
            return
        }

        val entity = userDAO.getUserByEmail(email)
        if (entity.isEmpty)
            showInvalidEmailMessage()
    }

    private def showInvalidEmailMessage() {
        form.recordError("No user associated with that e-mail.")
    }

    @CommitAfter
    def onSuccessFromForm() = {
        val user = userDAO.getUserByEmail(email)
        val token = new PasswordResetTokenEntity()
        token.setUser(user.get)
        database.save(token)

        val link = linkSource.createPageRenderLinkWithContext(classOf[ChangePasswordByTokenUser], token.getToken())
            .toAbsoluteURI(true)

        emailNotify.startMessage(EmailTemplate.USER_RESET_PASSWORD)
        emailNotify.setToAddress(email)
        emailNotify.addAttribute("axlink", link)
        emailNotify.send()

        classOf[PasswordResetSentUser]
    }
}
